File: Geeks For Geeks Solutions/Nodes At Given Distance In Binary Tree/PralinKhaira-Nodes At Given Distance In Binary Tree.cpp
/*
Problem Link: https://practice.geeksforgeeks.org/problems/nodes-at-given-distance-in-binary-tree/1
*/

/* A binary Tree node
struct Node
{
    int data;
    struct Node *left, *right;
};
*/

class Solution
{
private:

void markParent(Node* root, unordered_map<Node*, Node*>&parent_track){
        queue<Node*>q;
        q.push(root);
        while(!q.empty()){
            Node* node=q.front();
            q.pop();
            if(node->left){
                parent_track[node->left]=node;
                q.push(node->left);
            }
            if(node->right){
                parent_track[node->right]=node;
                q.push(node->right);
            }
        }
    }
public:
    Node* getRef(Node* root, int target)
    {
        if(root==NULL || root->data == target)
            return root;
        Node *left= getRef(root->left,target);
        Node *right= getRef(root->right, target);
        if(left!=NULL)
        return left;
        if(right!=NULL)
        return right;
    }
    
    vector <int> KDistanceNodes(Node* root, int target , int k)
    {
        // return the sorted vector of all nodes at k dist
        //Node* t=findRootOfTarget(root, target);
        Node* t= getRef(root,target);
        unordered_map<Node*, Node*>parent_track;
        markParent(root, parent_track);
        
        queue<Node*>q;
        unordered_map<Node*, bool>visited;
        q.push(t);
        visited[t]=1;
        int cnt=0;
        while(!q.empty()){
            int size=q.size();
            if(cnt++==k) break;
            for(int i=0; i<size; i++){
                Node* node=q.front();
                q.pop();
                if(node->left && !visited[node->left]){
                    q.push(node->left);
                    visited[node->left]=1;
                }
                if(node->right && !visited[node->right]){
                    q.push(node->right);
                    visited[node->right]=1;
                }
                if(parent_track[node] && !visited[parent_track[node]]){
                    q.push(parent_track[node]);
                    visited[parent_track[node]]=1;
                }
            }
        }
        vector<int>ans;
        while(!q.empty()){
            Node* cur=q.front();
            q.pop();
            ans.push_back(cur->data);
        }
        sort(ans.begin(), ans.end());
        return ans;
    }
};

File: Geeks For Geeks Solutions/Nodes At Given Distance In Binary Tree/PralinKhaira-Nodes At Given Distance In Binary Tree.py
# Problem Link: https://practice.geeksforgeeks.org/problems/nodes-at-given-distance-in-binary-tree/1

#User function Template for python3

class Node:
    def __init__(self, val):
        self.right = None
        self.data = val
        self.left = None
from collections import deque 
class Solution:
    
    def KDistanceNodes(self,root,target,k):
        #  relating the parents to their child nodes so that we can acces the path 
        dct = {}
        q = deque([root])
        while q:
            for i in range(len(q)):
                node = q.popleft()  # popping out the first element
                if node.data == target:
                    Target = node
                if node.left:   # if the left child exist 
                    dct[node.left] = node   # adding the child--> parent
                    q.append(node.left)   # appending the child into the queue
                if node.right:     # same if the right child exist 
                    dct[node.right] = node
                    q.append(node.right)
        
        # Step2 : 
        vis = set()
        pq = deque([Target])
        ds = 0
        while ds != k:
            for i in range(len(pq)):
                node = pq.popleft()
                vis.add(node)
                if node.left not in vis and node.left:
                    pq.append(node.left)
                    vis.add(node.left)
                if node.right not in vis and node.right:
                    pq.append(node.right)
                    vis.add(node.right)
                if dct.get(node)!=None and dct[node] not in vis:
                    pq.append(dct[node])
                    vis.add(dct[node])
            ds += 1
        ans = []
        while pq:
            ans.append(pq.pop().data)
        ans.sort()
        return ans
